library/core/src/mem/maybe_uninit.rs

```diff
@@ -298,6 +298,32 @@ impl<T> MaybeUninit<T> {
         MaybeUninit { value: ManuallyDrop::new(val) }
     }
 
+    /// Converts a mutable reference into a `MaybeUninit<T>` reference.
+    /// It is safe to call [`assume_init_drop`] on the return value of this function.
+    ///
+    /// Note that writing to a `MaybeUninit<T>` will never call `T`'s drop code.
+    /// It is your responsibility to make sure `T` gets dropped before writing a new value.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #![feature(maybe_uninit_from_mut)]
+    /// use std::mem::MaybeUninit;
+    ///
+    /// let mut v: Vec<u8> = vec![1, 2, 3, 4];
+    /// let uninit = MaybeUninit::from_mut(&mut v);
+    /// ```
+    #[unstable(feature = "maybe_uninit_from_mut", issue = "none")]
+    #[rustc_const_unstable(feature = "maybe_uninit_from_mut", issue = "none")]
+    #[must_use = "this conversion does nothing unless it's used"]
+    #[inline(always)]
+    pub const fn from_mut(val: &mut T) -> &mut MaybeUninit<T> {
```

Comment:
This is not sound or safe. You can un-initialize the value and then read it:
```rust
let mut initialized = String::from("ajahjjahjsjc;ejhc;je;jevj");
let maybe_uninit = MaybeUninit::from_mut(&mut initialized);
*maybe_uninit = MaybeUninit::uninit();

dbg!(initialized); // oh no
```
([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5bc1f334fd3734b6562ce4b02b357007))

Reply:
Reading the PR description that mentions `Drop`... maybe you've meant to add the method for [`ManuallyDrop`](https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html)?

Reply:
Oh, right, stacked borrows.

My brain was thinking that since the reference was moved, you could only use it via `MaybeUninit` going forward, but that's not true. There is probably a way of annotating the lifetimes to make that happen, but I'm not going to bother with it.

Will close.

Reply:
> Reading the PR description that mentions `Drop`... maybe you've meant to add the method for [`ManuallyDrop`](https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html)?

So actually... no. My main thought process was I could avoid having to make shims over the `write` method if I could simply return a reference to `MaybeUninit`. The problem is that, like mentioned, you can "undo" the conversion with stacked borrows. So, even though the actual `MaybeUninit` conversion is safe and sound, the fact that you can just stop assuming it's maybe-uninit at some point and then access the init reference directly means that it's not safe or sound.

Reply:
> There is probably a way of annotating the lifetimes to make that happen

I don't think there is, you can always reborrow a reference for a shorter lifetime, and then check the original reference. Well, the only way to avoid this is to accept `&'static mut T`, so that it can't be shorter than anything, but in this context it's not really useful.